core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala

```diff
@@ -0,0 +1,401 @@
+          val taskInfo = taskEnd.taskInfo
+          assert(taskInfo.index <= 1, s"Unknown task index ${taskInfo.index}")
+          if (taskInfo.index == 0) {
+            if (task0Killed.compareAndSet(false, true)) {
+              assert(taskInfo.successful)
+              // Since this task hasn't been killed before, it should still be at its first attempt.
+              assert(taskInfo.attemptNumber === 0)
+              val workerInfo = executorIdToWorkerInfo(taskInfo.executorId)
+              decommissionWorkerOnMaster(workerInfo, "Kill early done map worker")
+              killWorkerAfterTimeout(workerInfo, 0)
+              logInfo(s"Killed the node ${workerInfo.hostPort} that was running the early task")
+            } else {
+              assert(taskInfo.successful)
+              // The first attempt of this task should have failed since it was killed. Thus,
+              // either the task attempt or the stage attempt number should be more than 0.
+              assert(taskInfo.attemptNumber > 0 || taskEnd.stageAttemptId > 0)
+            }
+          } else {
+            // The second task is never touched and thus should succeed on the first attempt.
+            assert(taskInfo.successful)
+            assert(taskInfo.attemptNumber === 0)
+          }
+        }
+      }
+      sc.addSparkListener(listener)
+      val jobResult = sc.parallelize(1 to 2, 2).mapPartitionsWithIndex((pid, iter) => {
+        val sleepTimeSeconds = if (pid == 0) 1 else 10
+        Thread.sleep(sleepTimeSeconds * 1000L)
+        List.fill(pid + 1)(pid * 2 + 1).iterator
+      }, preservesPartitioning = true).repartition(1).sum()
+      assert(jobResult > 1)
+      // 4 tasks: 2 from first stage, one retry due to decom, one more in the second stage.
+      val tasksSeen = listener.getTasksFinished(10.seconds)
+      assert(tasksSeen.size >= 4, s"Expected at least 4 tasks but got $tasksSeen")
+    } finally {
+      ss.close()
+    }
+  }
+
+  test("decommission workers ensure that fetch failures lead to rerun") {
+    val conf = appConf
+    conf.set(config.Tests.TEST_NO_STAGE_RETRY, false)
+    conf.set(config.UNREGISTER_OUTPUT_ON_HOST_ON_FETCH_FAILURE, true)
+    makeWorkers(2)
+    sc = createSparkContext(conf)
+    val executorIdToWorkerInfo = getExecutorToWorkerAssignments
+    val executorToDecom = executorIdToWorkerInfo.keysIterator.next
+    val workerToDecomInfo = executorIdToWorkerInfo(executorToDecom)
+    val workerToDecomHost = workerToDecomInfo.host
+    val workerToDecomPort = workerToDecomInfo.port
+    // The setup of this job is similar to the one above: 2 stage job with first stage having
+    // long and short tasks. Except that we want the shuffle output to be regenerated on a
+    // fetch failure instead of an executor lost. Since it is hard to "trigger a fetch failure",
+    // we manually raise the FetchFailed exception when the 2nd stage's task runs and require that
+    // fetch failure to trigger a recomputation.
```

Comment:
One idea for fetch failure (I have not tried it yet): from workers you can access the worker then you have the `workDirPath` the executors on the workers are using subdirs of this work dir path. If you can delete its content (or just the shuffle files) between the stages I assume you will have a proper fetch failed.

Reply:
I spend several hours on this and couldn't get it to work :-(.  I understand that getting a proper fetch failure would obviously be more close to actual scenario here, but I couldn't get it working.